HSIEO-5803: Util to Enable IPM system-wide

Beiming-Zhang opened the pull request:
Use this new branch (Stable-IPM) for future development work. Closing the main PR in isc/zpm repo.
* Enabling zpm command in namespaces that currently do not have IPM installed.
* Pending question: Do we want to install zpm-perforce and zpm-studio-ui together with IPM? I think we should install them separately (presumably in HS-specific use cases). But open to discussion.

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,7 +2383,59 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+ } Else {
+	Set ZPMXMLDir = ##class(%File).NormalizeFilename("ZPM.xml", ##class(%File).ManagerDirectory())
+ 	Write !, "InterSystems Package Manager(IPM) is not installed in this namespace.", !
+ 	Set tList(1) = "Enable IPM for this namespace"
```
Do not use `t` prefix for variables

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,7 +2383,59 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+ 	Do ##class(%IPM.Main).Shell(pArgs...) Quit
+ } Else {
+	Set ZPMXMLDir = ##class(%File).NormalizeFilename("ZPM.xml", ##class(%File).ManagerDirectory())
+ 	Write !, "InterSystems Package Manager(IPM) is not installed in this namespace.", !
```
I think instead we should write out help text indicating how zpm can be enabled. Can list out namespaces where zpm exists already. Should have a new `enable` shell command which can have prompts and also modifiers to run silently ( like `all-ns` for all namespaces and with no modifiers enable for just the current namespace).

isc-kiyer replied:
Note: For the `enable` command will need to support installing from local artifact as well as community repo so should check for local artifact first and use that if provided. If not, look at remote repo. Should have modifiers to force looking at local artifact only or force looking at remote repo only as well.

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,7 +2383,59 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+ If $System.CLS.IsMthd("%IPM.Main", "Shell") { 
+ 	Do ##class(%IPM.Main).Shell(pArgs...) Quit
+ } Else {
+	Set ZPMXMLDir = ##class(%File).NormalizeFilename("ZPM.xml", ##class(%File).ManagerDirectory())
```
How will we handle multiple alrtifacts for different IPM versions? I think the artifact name needs to include the version string (it already does when you install from the community repo so we should keep that).

isc-kiyer replied:
Nit: Should ideally have a new folder /ipm under the manager directory for all the IPM artifacts

isc-kiyer reviewed (changes requested):
My comments are up. Need to think about this a bit more in terms of what enabling IPM can look like in difference scenarios + supporting different versions

Beiming-Zhang commented:
@isc-kiyer  Have updated the review with your requirements addressed mostly. One different opinion I have is that I think we should still enable IPM via zpm command directly issued in any namespace that does not yet have IPM. But it can have less customization comparing to the "enable" shell command. We can discuss that in detail if needed.

I've put some small concerns in "TODO" in this PR. Changes in this PR are aligned with what I've discussed with Michelle including the /Mgr/IPM/ file dir to store local artifact.

Please provide some feedback, and I will test it out meanwhile. Thank you!

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+	Set targetVersion = "" 		// var to store the final version of IPM to be installed
+	Kill targetNamespaces 		// multi-dim array to store the final namespaces that need to install IPM
+
+	// Get list of local IPM artifacts under /<instance install dir>/lib/IPM/
```
I recommend using all lower case so IPM --> ipm, even with kitting for releng

isc-kiyer commented on src/%IPM/Repo/Remote/PackageService.cls:
```diff
@@ -134,6 +134,22 @@ Method GetModuleManifest(pModuleReference As %IPM.Storage.ModuleInfo) As %Stream
+    Return tRequest.HttpResponse.Data
+  }
+
+  $$$ThrowStatus($$$ERROR($$$GeneralError,"Download package installer manifest, something went wrong."))
```
Should indicate module name and version in error message

isc-kiyer commented on src/%IPM/Repo/Remote/PackageService.cls:
```diff
@@ -134,6 +134,22 @@ Method GetModuleManifest(pModuleReference As %IPM.Storage.ModuleInfo) As %Stream
+  Set version = pModuleReference.VersionString
+
+  Set tSC = tRequest.Get($$$URLENCODE(tRequest.Location_"packages/" _ name _ "/"_version_"/installer"))
+  If $$$ISOK(tSC), tRequest.HttpResponse.StatusCode=200 {
```
Use && instead of ,

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -581,6 +581,25 @@ generate /my/path -export 00000,PacketName2,IgnorePacket2^00000,PacketName3,Igno
+		<!-- Modifiers -->
+		<modifier name="version" aliases="v" value="true" description="Specific version of IPM to be installed. Must specify if quiet flag is set."/>
+		<modifier name="namespaces" aliases="ns" value="true" description="Comma-separated namespaces in which IPM needs to be enabled."/>
+		<modifier name="globally" description="Will install IPM in all non-%SYS explicit namespaces that currently do not have IPM installed."/>
```
Since we have this globally enable option, please create corresponding jiras for new commands `disable` and and `ls` (to list IPM versions across namespaces)

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+		Set fileName = resultSet.%Get("Name")
+		If (resultSet.%Get("Type") = "F") {
+			// get the exact version string from filename
+			Set ipmLocalArray($Piece($Piece(fileName, "ipm-", 2), ".xml")) = fileName
```
Nit: Use 2, * in case ipm is part of version string in pre-release info

isc-kiyer replied:
This needs to still be addressed + also should use 1 to *-1 for the extension to be safe so the full expression should instead be the following:
```
$Piece($Piece(fileName, "ipm-", 2, *), ".xml", 1, *-1)
```

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+			// get the exact version string from filename
+			Set ipmLocalArray($Piece($Piece(fileName, "ipm-", 2), ".xml")) = fileName
+			Set counter = counter + 1
+			Set menuList(counter) = $Piece($Piece(fileName, "ipm-", 2), ".xml")
```
Nit: Instead of having a separate counter, can do `Set menuList ($Increment(menuList))`

isc-kiyer replied:
Nit: Looks like same $piece expression is evaluated twice. Place in a local variable so it is evaluated just once

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+	$$$ThrowOnError(sc)
+
+	If (localOnly && (counter = 0)) {
+		$$$ThrowOnError($$$ERROR($$$GeneralError,"No ipm-<version>.xml installer file is found locally."))
```
Indicate which directory you looked in

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+			If localOnly {
+				$$$ThrowOnError($$$ERROR($$$GeneralError,"No ipm-"_version_".xml installer file is found in "_XMLDir_" locally."))
+			} ElseIf 'quiet {
+				Write !, "No ipm-"_version_".xml installer file is found in "_XMLDir_" locally. Trying to download from the registry."
```
Should display this message either way. quiet only means user shouldn't be prompted. 
If you want a modifier for not displaying any messages to the current device, add a `verbose` modifier.

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+		If quiet {
+			$$$ThrowOnError($$$ERROR($$$GeneralError,"Cannot install IPM in non-prompt mode without a specified IPM version."))
+		}
+		If (counter = 0) {
```
I think this should just install the latest available version from the registry instead of throwing an error. Perhaps discuss with Michelle.

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+			$$$ThrowOnError($$$ERROR($$$GeneralError,"No IPM version is found in "_XMLDir_" locally. Use version modifier to specify a version to download from the registry."))
+		}
+		// prompt user with local options
+		Set menuValue = 1
```
The prompt should include local and remote versions for the user to choose from (need to accumulate available remote versions before hand)

isc-bzhang replied:
I'm somewhat hesitate to include remote versions in prompts, since they will be something look like ["0.0.4","0.0.5","0.0.6","0.0.7","0.1.0","0.1.1","0.1.2","0.1.3","0.1.4","0.1.5","0.1.6","0.1.7","0.1.8","0.2.1","0.2.10","0.2.11","0.2.12","0.2.13","0.2.14","0.2.2","0.2.3","0.2.4","0.2.5","0.2.6","0.2.7","0.2.8","0.2.9","0.3.0","0.3.2","0.3.3","0.3.4"] which is a lot for any sort of prompt. We can either let the user input a number for remote version in prompt if they really want to, or force to specify one as modifier. Or, provide only a "latest" option. I'm more leaning towards the last actually, but open to discuss

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+	
+	// Now, figure out which namespaces to install
+	If (quiet && (namespaces = "") && 'globally) {
+		$$$ThrowOnError($$$ERROR($$$GeneralError,"Cannot install IPM in non-prompt mode without specifying namespaces to install."))
```
If no namespaces provided, should just install in current namespace.

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
 
+ClassMethod EnableIPM(ByRef pCommandInfo)
+{
+	Set quiet = $$$HasModifier(pCommandInfo,"quiet")
```
At the beginning, should print current IPM version in namespace

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+	If (quiet && (namespaces = "") && 'globally) {
+		$$$ThrowOnError($$$ERROR($$$GeneralError,"Cannot install IPM in non-prompt mode without specifying namespaces to install."))
+	}
+	If ((namespaces '= "") && globally) {
```
Put all such validation checks at the beginning so no processing is done unless valid modifiers provided

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+	// TODO: Better and more stable way to check whether IPM has been enabled for a namespace instead of checking IPM classes
+	Set baseNS = $Namespace
+	New $Namespace
+	Set $Namespace = "%SYS"
```
No need to switch to %SYS

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+	$$$ThrowOnError(sc)
+	Kill namespacesWithoutIPM		// multi-dim array to store all non-%SYS explicit namepaces that currently do not have IPM installed
+	Set currentNS=$ORDER(allNS(""))
+	While (currentNS '="") {
```
Nit: Cleaner with the following structure (only need a single $Order):
```
Set currrentNS = ""
While 1 {
  Set currentNS = $Order(allNS(currentNS))
  If (currentNS = "") {
    Quit
  }
  // other logic here
}
```

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+	// Get all namespaces that currently do not have IPM enabled
+
+	// TODO: Let user interactively choose which namespaces to install IPM
+	// TODO: Better and more stable way to check whether IPM has been enabled for a namespace instead of checking IPM classes
```
I think checking for the main class is fine

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+		If (currentNS '= "%SYS") && ($ZCONVERT(currentNS, "U") = currentNS) {
+			Set $Namespace = currentNS
+			If '$System.CLS.IsMthd("%IPM.Main", "Shell") {
+				Set namespacesWithoutIPM(currentNS) = 1
```
Should also accumulate an array of namespaces with IPM and the version of IPM in those namespaces.
Based on namespaces selected/globally, may need to also upgrade IPM in those namespaces to match the provided version

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+	} Else {
+		// TODO: Implement this part
+		Write !, "Currently do not support interactively choose namespaces to install IPM. Use -namespaces or -globally modifier."
+		Quit
```
Nit: Use return instead of quit to exit a method

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+		Set ptr = 0
+   		While $ListNext(specifiedNSs, ptr, value) {
+			If ('$DATA(namespacesWithoutIPM(value)) && 'quiet) {
+				Write !, value_" is either not a valid namespace to install IPM, or already has IPM installed. Skipped."
```
This will need to be updated to do a version check as well based on comment above. If versions also match, then no need to reinstall

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2405,178 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+		If $IsObject(server) {
+			Set packageService = server.GetPackageService()
+			Set ipmRef = ##class(%IPM.Storage.ModuleInfo).%New()
+			Set ipmRef.Name = "zpm"
```
Mark TODO here to make this ipm eventually

isc-kiyer reviewed (changes requested):
Comments are up. Looks great as an initial version!

Beiming-Zhang commented:
@isc-kiyer  Have addressed most of the comments; I put a thumb-up to all the resolved ones.
For upgrade IPM, I think might be cleaner to do them separately from this command (right now I feel like it is doing too much already); any thoughts?
Also have put a comment for prompting user for remote versions above.

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -581,6 +581,28 @@ generate /my/path -export 00000,PacketName2,IgnorePacket2^00000,PacketName3,Igno
+			Enable IPM in other namespaces.
+		</summary>
+		<!-- Modifiers -->
+		<modifier name="version" aliases="v" value="true" description="Specific version of IPM to be installed. Must specify if quiet flag is set."/>
```
Nit: Perhaps can mention that a special version of `latest` can be provided. Also, adjust the text to indicate that if not specified, the latest version from the registry will be installed (hence is not required if quiet flag is set)

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2408,208 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+ClassMethod EnableIPM(ByRef pCommandInfo)
+{
+	Write !, "Version of IPM in current namespace: "
+	Do ..GetListModules("zpm",.list)
```
Nit: Perhaps put the name of the module in a macro (and use the same macro in the `Version()` method of this class

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2408,208 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+	}
+  	Set statement = ##class(%SQL.Statement).%New()
+  	Set sc = statement.%PrepareClassQuery("%File", "FileSet")
+  	If $$$ISERR(sc) {
```
Nit: Can just do `$$$ThrowOnError(sc)`

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2408,208 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+	}
+	// Valid IPM installation manifest should be of format: ipm-0.0.1.xml
+	Set resultSet = statement.%Execute(XMLDir, "ipm-*.xml")
+	If ((resultSet.%SQLCODE '= 0) && '(resultSet.%Message [ "does not exist")) {
```
Nit: Here should check if %SQLCODE < 0 since you could get 100 I believe if there are no files found. Then no need to check for the does not exist error either and can directly use the `$$$ThrowSQLIfError` macro

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2408,208 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+		Set fileName = resultSet.%Get("Name")
+		If (resultSet.%Get("Type") = "F") {
+			// get the exact version string from filename
+			Set exactIPMVer = $Piece($Piece(fileName, "ipm-", 2, *), ".xml")
```
Nit: For same reasoning in case `.xml` is for some bad reason in the pre-release info, make the second $Piece the following: `..., ".xml", 1, *-1)`

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2408,208 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+
+	If (version '= "") {
+		Set targetVersion = version
+		If '$DATA(ipmLocalArray(version)) {
```
Hmm can you create a Jira for deciding whether we want to support always obtaining a version from the registry? Right now if a local version is available, we are always using that but may want to force getting from registry.
In future, once packages have hashes (not sure if they do), we could do a hash check against the registry and local versions to confirm they are the same.

isc-bzhang replied:
Sounds good, will create a jira for it and describe the current settings we have (in which case do local, in which case do remote, etc.)
https://usjira.iscinternal.com/browse/HSIEO-6971

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2408,208 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+					Set packageService = server.GetPackageService()
+					Set jsonStream = packageService.GetModuleData("zpm")
+					Set versions = {}.%FromJSON(jsonStream)
+					// TO DECIDE: add all those versions to menuList and prompt user? Or...?
```
I think just add the latest version here

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2408,208 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+		If currentNS = "" {
+			Quit
+		}
+		// avoid %SYS and implicit namespaces
```
Nit: `implied` not `implicit`

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2408,208 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+		Set currentNS=$ORDER(targetNamespaces(""))
+		While (currentNS '="") {
+			Set $Namespace = currentNS
+			Do $SYSTEM.OBJ.Load(##class(%File).NormalizeFilename(targetIPMFileName, XMLDir), "ck")
```
Should accumulate status per namespace and only show IPM enabled message if was successful. Should abandon if IPM failed to install in any namespace instead of continuing.

isc-kiyer replied:
Also, use the `-d` flag so that minimal output is displayed (only see errors and not list of classes loaded)

isc-kiyer reviewed (changes requested):
Final round of comments

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -581,6 +581,32 @@ generate /my/path -export 00000,PacketName2,IgnorePacket2^00000,PacketName3,Igno
+		<!-- Modifiers -->
+		<modifier name="version" aliases="v" value="true" description="A special version of IPM can be provided. If not specified, the latest version from the registry will be installed (hence is not required if quiet flag is set)."/>
+		<modifier name="namespaces" aliases="ns" value="true" description="Comma-separated namespaces in which IPM needs to be enabled."/>
+		<modifier name="globally" description="Will install IPM in all non-%SYS explicit namespaces that currently do not have IPM installed."/>
```
For all these boolean modifiers, indicate default is false.

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -723,16 +749,19 @@ ClassMethod ShellInternal(pCommand As %String, Output pException As %Exception.A
+				Set pException = $$$NULLOREF
+				Return
 			}
+        	Do ..DisplayError(pException.AsStatus())
```
Nit: Fix indentation

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2412,229 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+
+	// basic sanity checks
+	If (quiet && (namespaces = "") && 'globally) {
+		$$$ThrowOnError($$$ERROR($$$GeneralError,"Cannot install IPM in non-prompt mode without specifying namespaces to install."))
```
Nit: Since flag is called quiet, can call it quiet mode instead of non-prompt mode.

isc-kiyer commented on src/%IPM/Repo/Remote/PackageService.cls:
```diff
@@ -20,6 +22,14 @@ Method GetInfo(url As %String) As %DynamicObject
 	$$$ThrowStatus($$$ERROR($$$GeneralError,"Registry server not available."))
 }
 
+Method GetLatestIPMVersion() As %String
```
Nit: These methods should not be for a specific module ideally so perhaps just change it to `GetLatestModuleVersion` and pass as an argument the module name

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2412,229 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+		If quiet {
+			// Non-prompt mode + no version specified - either error out or install latest version from registry
+			If localOnly {
+				$$$ThrowOnError($$$ERROR($$$GeneralError,"To install local IPM in non-prompt mode, a specific version must be specified using the version modifier."))
```
Why not just install the highest available version locally? You have all the versions in the multidim array so can use semantic version comparison functions to figure out the highest version.

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2412,229 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+			}
+		} Else {
+			// prompt user with option to download the latest version from registry
+			Set localIPMCount = menuList
```
Nit: Should be `$Get(menuList,0)`

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2412,229 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+				$$$ThrowOnError($$$ERROR($$$GeneralError,"Unsupported response. Operation cancelled."))
+			}
+			Set targetVersion = $Piece(menuList(menuValue), " ") // get rid of (local version) or (remote version) suffix
+			zw targetVersion
```
Should not use zw. Should use a clear write statement if you want to write out the version selected.

isc-bzhang replied:
Ah this is debugging zw, now removed

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2412,229 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+			}
+			// avoid %SYS and implied namespaces
+			If (currentNS '= "%SYS") && ($ZCONVERT(currentNS, "U") = currentNS) {
+				New $Namespace
```
Should have a single new $Namespace outside the loop instead of in here.

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2412,229 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+				}
+				New $Namespace
+				Set $Namespace = currentNS
+				Do $SYSTEM.OBJ.LoadStream(manifest, "ck/displayerror=0/displaylog=1")
```
When loading, should only displayerror and not displaylog. Should use same flags for LoadStream here and Load below for local version so perhaps store flags in a local variable.

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2412,229 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+				If currentNS = "" {
+					Quit
+				}
+				New $Namespace
```
Again the New $Namespace isn't needed since it has already been done earlier. I recommend moving it to the beginning of the method itself.

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2412,229 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+			Set $Namespace = currentNS
+			Set tSC = $SYSTEM.OBJ.Load(##class(%File).NormalizeFilename(targetIPMFileName, XMLDir), "ck")
+			If $$$ISOK(tSC) {
+				Set enabledNSList = enabledNSList_$LB(currentNS)
```
Nit: Use full form of commands ($ListBuild)

isc-kiyer commented on src/%IPM/Main.cls:
```diff
@@ -2383,11 +2412,229 @@ ClassMethod UpdateLanguageExtensions(pVerbose As %Boolean = 0, pTestOnly As %Boo
+			} else {
+				Write !, "IPM enabled for namespace(s) "_$ListToString(enabledNSList)
+				Write !, "IPM failed to enable for namespace "_currentNS
+				$$$ThrowOnError(tSC)
```
Rather than throwing error, I suggest accumulating the errors and continuing so that IPM gets enabled in as many namespaces as possible. Then outside the loop, can indicate which namespaces it was successfully enabled/upgrade for and which ones it failed for. The same printing logic can be used for whether using remote or local artifact so can be outside the if-else block

isc-kiyer reviewed (changes requested):
@Beiming-Zhang Few final comments. Looks great!

isc-kiyer reviewed (approved):
Looks great!
